flycheck.el

```diff
@@ -4095,19 +4140,29 @@ refresh the mode line."
 
 STATUS defaults to `flycheck-last-status-change' if omitted or
 nil."
-  (let ((text (pcase (or status flycheck-last-status-change)
-                (`not-checked "")
-                (`no-checker "-")
-                (`running "*")
-                (`errored "!")
-                (`finished
-                 (let-alist (flycheck-count-errors flycheck-current-errors)
-                   (if (or .error .warning)
-                       (format ":%s|%s" (or .error 0) (or .warning 0))
-                     "")))
-                (`interrupted ".")
-                (`suspicious "?"))))
-    (concat " " flycheck-mode-line-prefix text)))
+  (let (text
+        face)
+    (setq text (concat " " flycheck-mode-line-prefix
+                       (pcase (or status flycheck-last-status-change)
+                         (`not-checked "")
+                         (`no-checker "-")
+                         (`running "*")
+                         (`errored (progn
+                                     (setq face 'error)
+                                     "!"))
+                         (`finished
+                          (let-alist (flycheck-count-errors flycheck-current-errors)
+                            (if (or .error .warning)
+                                (progn
+                                  (setq face 'error)
+                                  (format ":%s|%s" (or .error 0) (or .warning 0)))
+                              (setq face 'success)
+                              ":0")))
+                         (`interrupted ".")
+                         (`suspicious "?"))))
+    (when (and face flycheck-mode-line-color)
```

Comment:
I'm guessing you can just use `if/else` here and have the normal and colorized version of the text as the two branches. (or something along those lines) I'd also suggest using `format` over `concat`.

Reply:
Separated the mode indicator text from the face selection